pkg/server/plugin/nodeattestor/httpchallenge/httpchallenge.go

```diff
@@ -0,0 +1,272 @@
+	configv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/service/common/config/v1"
+	"github.com/spiffe/spire/pkg/common/agentpathtemplate"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/common/plugin/httpchallenge"
+	nodeattestorbase "github.com/spiffe/spire/pkg/server/plugin/nodeattestor/base"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	pluginName = "http_challenge"
+)
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.BuiltIn {
+	return catalog.MakeBuiltIn(pluginName,
+		nodeattestorv1.NodeAttestorPluginServer(p),
+		configv1.ConfigServiceServer(p),
+	)
+}
+
+type configuration struct {
+	trustDomain       spiffeid.TrustDomain
+	pathTemplate      *agentpathtemplate.Template
+	requiredPort      *int
+	allowNonRootPorts bool
+	dnsPatterns       []*regexp.Regexp
+	agentNamePattern  *regexp.Regexp
+	tofu              bool
+}
+
+type Config struct {
+	DNSPatterns       []string `hcl:"dns_patterns"`
+	RequiredPort      *int     `hcl:"required_port"`
+	AllowNonRootPorts *bool    `hcl:"allow_non_root_ports"`
+	AgentPathTemplate string   `hcl:"agent_path_template"`
+	TOFU              *bool    `hcl:"tofu"`
+}
+
+type Plugin struct {
+	nodeattestorbase.Base
+	nodeattestorv1.UnsafeNodeAttestorServer
+	configv1.UnsafeConfigServer
+
+	m      sync.Mutex
+	config *configuration
+
+	log hclog.Logger
+}
+
+func New() *Plugin {
+	return &Plugin{}
+}
+
+func (p *Plugin) Attest(stream nodeattestorv1.NodeAttestor_AttestServer) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	payload := req.GetPayload()
+	if payload == nil {
+		return status.Error(codes.InvalidArgument, "missing attestation payload")
+	}
+
+	attestationData := new(httpchallenge.AttestationData)
+	if err := json.Unmarshal(payload, attestationData); err != nil {
+		return status.Errorf(codes.InvalidArgument, "failed to unmarshal data: %v", err)
+	}
+
+	if config.requiredPort != nil && attestationData.Port != *config.requiredPort {
+		return status.Errorf(codes.InvalidArgument, "port %d is not allowed to be used by this server", attestationData.Port)
+	}
+	if (!config.allowNonRootPorts) && attestationData.Port >= 1024 {
+		return status.Errorf(codes.InvalidArgument, "port %d is not allowed to be >= 1024", attestationData.Port)
+	}
+
+	l := config.agentNamePattern.FindAllStringSubmatch(attestationData.AgentName, -1)
+	if len(l) != 1 || len(l[0]) == 0 || len(l[0]) > 32 {
+		return status.Error(codes.InvalidArgument, "agent name is not valid")
+	}
+
+	notfound := false
+	for _, re := range config.dnsPatterns {
+		notfound = true
+		l := re.FindAllStringSubmatch(attestationData.HostName, -1)
+		if len(l) > 0 {
+			notfound = false
+			break
+		}
+	}
+	if notfound {
+		return status.Errorf(codes.PermissionDenied, "the requested hostname is not allowed to connect")
+	}
```

Comment:
nit: I think it would be cleaner to extract this to a function. The notFound variable wouldn't be needed then (function could early-return if the hostname matches a pattern).